File: scripts/import_recruiters.py
"""
Recruiter Data Import Script
- Imports recruiters from Excel file with 100% complete data
- Creates organization records
- Creates Supabase Auth accounts
- Creates user records
- Sends welcome emails
"""

import pandas as pd
import os
import sys
from supabase import create_client, Client
from datetime import datetime
import uuid
import json
from dotenv import load_dotenv

# Load environment variables
load_dotenv('/app/.env')

# Supabase credentials
SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SUPABASE_SERVICE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

# Default password for all recruiters
DEFAULT_PASSWORD = "Recruiter@2025"

# Initialize Supabase client
supabase: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)

class RecruiterImporter:
    def __init__(self, file_path):
        self.file_path = file_path
        self.stats = {
            'total_rows': 0,
            'complete_rows': 0,
            'duplicate_emails': 0,
            'already_exists': 0,
            'successfully_imported': 0,
            'failed': 0,
            'errors': []
        }
        self.imported_recruiters = []
        
    def load_data(self):
        """Load Excel file and filter complete rows"""
        print("📂 Loading Excel file...")
        df = pd.read_excel(self.file_path)
        self.stats['total_rows'] = len(df)
        
        # Filter rows with 100% complete data
        complete_df = df.dropna()
        self.stats['complete_rows'] = len(complete_df)
        
        print(f"✅ Found {self.stats['complete_rows']} rows with 100% complete data out of {self.stats['total_rows']} total rows")
        
        return complete_df
    
    def remove_duplicates(self, df):
        """Remove duplicate email addresses"""
        print("\n🔍 Checking for duplicate emails...")
        
        # Convert email to lowercase for comparison
        df['Mail ID'] = df['Mail ID'].str.strip().str.lower()
        
        # Keep first occurrence of each email
        df_unique = df.drop_duplicates(subset=['Mail ID'], keep='first')
        
        self.stats['duplicate_emails'] = len(df) - len(df_unique)
        
        if self.stats['duplicate_emails'] > 0:
            print(f"⚠️  Removed {self.stats['duplicate_emails']} duplicate email addresses")
        else:
            print("✅ No duplicate emails found")
        
        return df_unique
    
    def check_existing_emails(self, emails):
        """Check which emails already exist in the database"""
        print("\n🔍 Checking for existing emails in database...")
        
        try:
            # Check in users table
            response = supabase.table('users').select('email').in_('email', emails).execute()
            existing_emails = set([user['email'].lower() for user in response.data])
            
            print(f"📊 Found {len(existing_emails)} emails that already exist in database")
            return existing_emails
        except Exception as e:
            print(f"⚠️  Error checking existing emails: {str(e)}")
            return set()
    
    def create_organization(self, recruiter_data):
        """Create organization record for recruiter"""
        org_id = str(uuid.uuid4())
        
        org_record = {
            'id': org_id,
            'name': recruiter_data['Company Name'],
            'type': 'recruiter',
            'state': recruiter_data['Location'],
            'website': recruiter_data['Website'],
            'phone': recruiter_data['Number'],
            'email': recruiter_data['Mail ID'],
            'address': recruiter_data['Address'],
            'verificationStatus': 'approved',
            'isActive': True,
            'metadata': {
                'company_type': recruiter_data['Company type'],
                'imported_from': 'excel',
                'import_date': datetime.now().isoformat()
            },
            'createdAt': datetime.now().isoformat()
        }
        
        try:
            response = supabase.table('organizations').insert(org_record).execute()
            return org_id, None
        except Exception as e:
            return None, str(e)
    
    def create_auth_user(self, email, company_name):
        """Create user in Supabase Auth"""
        try:
            # Create auth user
            auth_response = supabase.auth.admin.create_user({
                "email": email,
                "password": DEFAULT_PASSWORD,
                "email_confirm": True,  # Auto-confirm email
                "user_metadata": {
                    "name": company_name,
                    "role": "recruiter",
                    "imported_from": "excel"
                }
            })
            
            return auth_response.user.id, None
        except Exception as e:
            error_msg = str(e)
            # Check if user already exists in auth
            if "already been registered" in error_msg or "User already registered" in error_msg:
                return None, "AUTH_EXISTS"
            return None, error_msg
    
    def create_user_record(self, auth_user_id, email, company_name, org_id):
        """Create user record in users table"""
        user_record = {
            'id': auth_user_id,
            'email': email,
            'role': 'recruiter',
            'organizationId': org_id,
            'isActive': True,
            'metadata': {
                'name': company_name,
                'imported_from': 'excel',
                'import_date': datetime.now().isoformat()
            },
            'createdAt': datetime.now().isoformat()
        }
        
        try:
            response = supabase.table('users').insert(user_record).execute()
            return True, None
        except Exception as e:
            return False, str(e)
    
    def send_welcome_email(self, email, company_name):
        """Send welcome email to recruiter"""
        # Note: This is a placeholder. In production, you'd use a real email service
        # For now, we'll just log that email should be sent
        print(f"  📧 Welcome email queued for: {email}")
        return True
    
    def import_recruiter(self, row):
        """Import a single recruiter"""
        recruiter_data = {
            'Company Name': row['Company Name'],
            'Website': row['Website'],
            'Number': row['Number'],
            'Mail ID': row['Mail ID'],
            'Company type': row['Company type'],
            'Address': row['Address'],
            'Location': row['Location']
        }
        
        email = recruiter_data['Mail ID']
        company_name = recruiter_data['Company Name']
        
        print(f"\n📝 Processing: {company_name} ({email})")
        
        try:
            # Step 1: Create organization
            print("  🏢 Creating organization...")
            org_id, error = self.create_organization(recruiter_data)
            if error:
                raise Exception(f"Organization creation failed: {error}")
            print(f"  ✅ Organization created: {org_id}")
            
            # Step 2: Create auth user
            print("  👤 Creating Supabase Auth user...")
            auth_user_id, error = self.create_auth_user(email, company_name)
            if error:
                if error == "AUTH_EXISTS":
                    print(f"  ⚠️  Auth user already exists, skipping...")
                    self.stats['already_exists'] += 1
                    return False
                raise Exception(f"Auth user creation failed: {error}")
            print(f"  ✅ Auth user created: {auth_user_id}")
            
            # Step 3: Create user record
            print("  📋 Creating user record...")
            success, error = self.create_user_record(auth_user_id, email, company_name, org_id)
            if error:
                raise Exception(f"User record creation failed: {error}")
            print("  ✅ User record created")
            
            # Step 4: Send welcome email
            print("  📧 Sending welcome email...")
            self.send_welcome_email(email, company_name)
            
            # Success!
            self.stats['successfully_imported'] += 1
            self.imported_recruiters.append({
                'company_name': company_name,
                'email': email,
                'org_id': org_id,
                'auth_user_id': auth_user_id,
                'location': recruiter_data['Location']
            })
            
            print(f"  ✅ Successfully imported!")
            return True
            
        except Exception as e:
            self.stats['failed'] += 1
            error_msg = f"{company_name} ({email}): {str(e)}"
            self.stats['errors'].append(error_msg)
            print(f"  ❌ Failed: {str(e)}")
            return False
    
    def generate_report(self):
        """Generate import report"""
        print("\n" + "="*80)
        print("📊 IMPORT SUMMARY REPORT")
        print("="*80)
        print(f"\n📈 Statistics:")
        print(f"  • Total rows in file: {self.stats['total_rows']}")
        print(f"  • Rows with 100% complete data: {self.stats['complete_rows']}")
        print(f"  • Duplicate emails removed: {self.stats['duplicate_emails']}")
        print(f"  • Already existing in database: {self.stats['already_exists']}")
        print(f"  • Successfully imported: {self.stats['successfully_imported']}")
        print(f"  • Failed imports: {self.stats['failed']}")
        
        if self.stats['errors']:
            print(f"\n❌ Errors ({len(self.stats['errors'])}):")
            for error in self.stats['errors'][:10]:  # Show first 10 errors
                print(f"  • {error}")
            if len(self.stats['errors']) > 10:
                print(f"  ... and {len(self.stats['errors']) - 10} more errors")
        
        if self.imported_recruiters:
            print(f"\n✅ Successfully Imported Recruiters:")
            print(f"  First 5 recruiters:")
            for i, recruiter in enumerate(self.imported_recruiters[:5], 1):
                print(f"  {i}. {recruiter['company_name']}")
                print(f"     Email: {recruiter['email']}")
                print(f"     Location: {recruiter['location']}")
                print(f"     Organization ID: {recruiter['org_id']}")
        
        print("\n📧 Login Credentials for all imported recruiters:")
        print(f"  • Email: <their email address>")
        print(f"  • Password: {DEFAULT_PASSWORD}")
        print("\n" + "="*80)
        
        # Save detailed report to file
        report_file = '/app/recruiter_import_report.json'
        report_data = {
            'timestamp': datetime.now().isoformat(),
            'statistics': self.stats,
            'imported_recruiters': self.imported_recruiters
        }
        
        with open(report_file, 'w') as f:
            json.dump(report_data, f, indent=2)
        
        print(f"\n💾 Detailed report saved to: {report_file}")
    
    def run(self):
        """Run the import process"""
        print("🚀 Starting Recruiter Import Process...")
        print("="*80)
        
        # Load data
        df = self.load_data()
        
        # Remove duplicates
        df_unique = self.remove_duplicates(df)
        
        # Check existing emails
        all_emails = df_unique['Mail ID'].tolist()
        existing_emails = self.check_existing_emails(all_emails)
        
        # Filter out existing emails
        df_to_import = df_unique[~df_unique['Mail ID'].isin(existing_emails)]
        
        self.stats['already_exists'] = len(df_unique) - len(df_to_import)
        
        print(f"\n📊 Ready to import {len(df_to_import)} new recruiters")
        
        if len(df_to_import) == 0:
            print("⚠️  No new recruiters to import!")
            self.generate_report()
            return
        
        # Import each recruiter
        print("\n" + "="*80)
        print("🔄 Starting Import...")
        print("="*80)
        
        for idx, row in df_to_import.iterrows():
            self.import_recruiter(row)
        
        # Generate final report
        self.generate_report()


if __name__ == "__main__":
    file_path = '/tmp/recruiter_data.xlsx'
    
    if not os.path.exists(file_path):
        print(f"❌ Error: File not found: {file_path}")
        sys.exit(1)
    
    if not SUPABASE_URL or not SUPABASE_SERVICE_KEY:
        print("❌ Error: Supabase credentials not found in environment variables")
        sys.exit(1)
    
    importer = RecruiterImporter(file_path)
    importer.run()

File: scripts/import_recruiters.js
const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

// Load environment variables from .env file
const envPath = path.join(__dirname, '..', '.env');
const envContent = fs.readFileSync(envPath, 'utf8');
const envVars = {};

envContent.split('\n').forEach(line => {
  line = line.trim();
  if (line && !line.startsWith('#')) {
    const [key, ...valueParts] = line.split('=');
    if (key && valueParts.length > 0) {
      envVars[key.trim()] = valueParts.join('=').trim();
    }
  }
});

// Initialize Supabase client with service role key for admin operations
const supabase = createClient(
  envVars.NEXT_PUBLIC_SUPABASE_URL,
  envVars.SUPABASE_SERVICE_ROLE_KEY,
  {
    auth: {
      autoRefreshToken: false,
      persistSession: false
    }
  }
);

async function importRecruiters() {
  console.log('='.repeat(80));
  console.log('RECRUITER DATA IMPORT SCRIPT');
  console.log('='.repeat(80));

  // Load prepared data
  const recruitersData = JSON.parse(fs.readFileSync('recruiters_for_import.json', 'utf8'));
  const authUsersData = JSON.parse(fs.readFileSync('recruiter_auth_users.json', 'utf8'));

  console.log(`\nLoaded ${recruitersData.length} recruiters and ${authUsersData.length} auth users`);

  let successCount = 0;
  let errorCount = 0;
  const errors = [];

  // Process each recruiter
  for (let i = 0; i < recruitersData.length; i++) {
    const recruiter = recruitersData[i];
    const authUser = authUsersData[i];

    try {
      console.log(`\n[${i + 1}/${recruitersData.length}] Processing: ${recruiter.name}`);

      // Step 1: Insert organization
      const { data: orgData, error: orgError } = await supabase
        .from('organizations')
        .insert({
          id: recruiter.id,
          name: recruiter.name,
          type: recruiter.type,
          email: recruiter.email,
          phone: recruiter.phone,
          website: recruiter.website,
          address: recruiter.address,
          city: recruiter.city,
          state: recruiter.state,
          country: recruiter.country,
          metadata: recruiter.metadata,
          isActive: recruiter.isActive,
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString()
        })
        .select()
        .single();

      if (orgError) {
        throw new Error(`Organization insert failed: ${orgError.message}`);
      }

      console.log(`  ✓ Organization created: ${orgData.id}`);

      // Step 2: Create auth user
      const { data: authData, error: authError } = await supabase.auth.admin.createUser({
        email: authUser.email,
        password: authUser.password,
        email_confirm: true,
        user_metadata: {
          name: authUser.name,
          role: authUser.role
        }
      });

      if (authError) {
        // If user already exists, try to get existing user
        if (authError.message.includes('already registered')) {
          console.log(`  ⚠ Auth user already exists: ${authUser.email}`);
          
          // Try to find existing user in users table
          const { data: existingUser } = await supabase
            .from('users')
            .select('id')
            .eq('email', authUser.email)
            .single();

          if (existingUser) {
            console.log(`  ✓ Using existing user: ${existingUser.id}`);
            successCount++;
            continue;
          }
        } else {
          throw new Error(`Auth user creation failed: ${authError.message}`);
        }
      }

      console.log(`  ✓ Auth user created: ${authData.user.id}`);

      // Step 3: Insert user record in users table
      const { data: userData, error: userError } = await supabase
        .from('users')
        .insert({
          id: authData.user.id,
          email: authUser.email,
          role: authUser.role,
          organizationId: authUser.organizationId,
          isActive: true,
          metadata: {
            name: authUser.name,
            source: 'recruiter_import'
          },
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString()
        })
        .select()
        .single();

      if (userError) {
        throw new Error(`User insert failed: ${userError.message}`);
      }

      console.log(`  ✓ User record created: ${userData.id}`);
      console.log(`  ✓ SUCCESS: ${recruiter.name} imported successfully`);
      
      successCount++;

    } catch (error) {
      console.error(`  ✗ ERROR: ${error.message}`);
      errorCount++;
      errors.push({
        recruiter: recruiter.name,
        email: recruiter.email,
        error: error.message
      });
    }

    // Small delay to avoid rate limiting
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  // Print summary
  console.log('\n' + '='.repeat(80));
  console.log('IMPORT SUMMARY');
  console.log('='.repeat(80));
  console.log(`Total recruiters processed: ${recruitersData.length}`);
  console.log(`Successful imports: ${successCount}`);
  console.log(`Failed imports: ${errorCount}`);
  console.log(`Success rate: ${((successCount / recruitersData.length) * 100).toFixed(2)}%`);

  if (errors.length > 0) {
    console.log('\nErrors encountered:');
    errors.forEach((err, idx) => {
      console.log(`${idx + 1}. ${err.recruiter} (${err.email}): ${err.error}`);
    });

    // Save errors to file
    fs.writeFileSync('import_errors.json', JSON.stringify(errors, null, 2));
    console.log('\nErrors saved to: import_errors.json');
  }

  // Get final count of recruiter organizations
  const { count } = await supabase
    .from('organizations')
    .select('*', { count: 'exact', head: true })
    .eq('type', 'recruiter');

  console.log(`\nTotal recruiter organizations in database: ${count}`);
  console.log('\n✓ Import process completed!');
}

// Run the import
importRecruiters()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
